docroot/sites/classrooms.uiowa.edu/modules/classrooms_core/src/RoomItemProcessor.php

```diff
@@ -0,0 +1,212 @@
+      }
+    }
+
+    // Mapping Room Features and Technology Features fields
+    // to the featureList value from endpoint.
+    if (isset($record->featureList)) {
+      $query = \Drupal::entityQuery('taxonomy_term')->orConditionGroup()
+        ->condition('vid', 'room_features')
+        ->condition('vid', 'technology_features');
+
+      $tids = \Drupal::entityQuery('taxonomy_term')
+        ->condition($query)
+        ->execute();
+      if ($tids) {
+        $storage = \Drupal::entityTypeManager()->getStorage('taxonomy_term');
+        $terms = $storage->loadMultiple($tids);
+        $room_features = [];
+        $tech_features = [];
+        foreach ($terms as $term) {
+          if ($api_mapping = $term->get('field_api_mapping')?->value) {
+            if (in_array($api_mapping, $record->featureList)) {
+              if ($term->bundle() === 'room_features') {
+                $room_features[] = $term->id();
+              }
+              else {
+                $tech_features[] = $term->id();
+              }
+            }
+          }
+        }
+        if (!empty($room_features)) {
+          // Cheat it a bit by fetching a string and exploding it
+          // to end up with a basic array of target ids.
+          $entity_features = $entity->get('field_room_features')->getString();
+          $entity_features = explode(', ', $entity_features);
+          // Sort lists before comparing.
+          sort($entity_features);
+          sort($room_features);
+          if ($entity_features !== $room_features) {
+            $updated = TRUE;
+            $entity->set('field_room_features', $room_features);
+          }
+        }
+
+        if (!empty($tech_features)) {
+          // Cheat it a bit by fetching a string and exploding it
+          // to end up with a basic array of target ids.
+          $entity_features = $entity->get('field_room_technology_features')->getString();
+          $entity_features = explode(', ', $entity_features);
+          // Sort lists before comparing.
+          sort($entity_features);
+          sort($tech_features);
+          if ($entity_features !== $tech_features) {
+            $updated = TRUE;
+            $entity->set('field_room_technology_features', $tech_features);
+          }
+        }
+      }
+    }
+    return $updated;
+  }
+
+  /**
+   * Returns a single record for the entity.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The entity to pull a record for.
+   *
+   * @return array|mixed
+   *   The record or an empty array.
+   */
+  public static function getRecord(ContentEntityInterface $entity): mixed {
+    // Grab MAUI room data.
+    $building_id = $entity
+      ->field_room_building_id
+      ?->target_id;
+    $room_id = $entity
+      ->field_room_room_id
+      ?->value;
+    /** @var \Drupal\uiowa_maui\MauiApi $maui_api */
+    $maui_api = \Drupal::service('uiowa_maui.api');
+    // @todo Need to handle $building_id or $room_id being NULL.
```

Comment:
Is this handled below? If so, should this be removed?

Reply:
I don't think this is currently handled.

https://github.com/uiowa/uiowa/blob/7f48860680cf1bb3d85d35a48e03a97bef5a361b/docroot/modules/custom/uiowa_maui/src/MauiApi.php#L297-L299

Possibly we need to handle it in the above snippet instead, but since it is not currently handled there, if either `$building_id` or `$room_id` was `NULL`, it would fail ungracefully.

Reply:
Shoot. It does error in the case of `building_id` being NULL and `room_id` being filled. As long as `room_id` is null (or null as well as `building_id`, it's a successful request that just doesn't end up with data, and so the other catch works fine.

I'd only checked the case if both were NULL.

Reply:
When `NULL` is (automatically) cast to a string, does it become `''`? Or `'null'`?